[BOLT][AArch64] Enabling Inlining for Memcpy for AArch64 in BOLT

yafet-a opened the pull request:
## Overview
The pass for inlining memcpy in BOLT was currently X86-specific. It was using the instruction `rep movsb` which currently has no equivalent in ARM v8-A.

This patch implements a static size analysis system for AArch64 memcpy inlining that extracts copy sizes from preceding instructions to then use it to generate the optimal width-specific load/store sequences.

## Testing Coverage (`inline-memcpy.s`)
### Positive Tests:
- [x] Exact size optimizations: 1, 2, 4, 8, 16, 32 bytes → optimal instruction sequences
- [x] Arbitrary size decomposition: 37 bytes → 16+16+4+1 byte sequence with correct offsets
- [x] Inline count verification: CHECK-INLINE: inlined 8 memcpy() calls
- [x] Assembly validation: CHECK-ASM patterns verify exact generated instructions
### Negative Tests:
- [x] Large size safety: 128 bytes → CHECK-ASM-NOT: ldr.*q (no SIMD, skipped inlining)
- [x] No unwanted instructions: CHECK-ASM-NOT patterns ensure clean generation

yafet-a committed:
pre-commit test

yafet-a committed:
[BOLT] documentation

yafet-a committed:
[BOLT][AArch64] Implement safe size-aware memcpy inlining

yafet-a committed:
test target fix for CI cross-compilation issue

yafet-a committed:
moved inline-memcpy to avoid CI cross-compilation PIE conflicts

yafet-a committed:
removed old test

sjoerdmeijer commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -2597,6 +2597,210 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+    switch (Size) {
+    case 1:
+      // Single byte copy
+      Code.emplace_back(MCInstBuilder(AArch64::LDRBBui)
```
All this code is duplicated here. The only thing that changes are the LD and STR opcodes. Maybe create a switch over `Size`, and just record the load/store opcodes, then have just 2 calls to create the instructions. I see a bit down that some registers might be different as well, but hopefully you'll get the gist of this and can reduce the code a bit.

yafet-a committed:
response to review

yafet-a commented on bolt/test/runtime/AArch64/inline-memcpy.s:
```diff
@@ -0,0 +1,193 @@
+
+# RUN: llvm-mc -filetype=obj -triple aarch64-unknown-unknown %s -o %t.o
+# RUN: %clang %cflags -no-pie %t.o -o %t.exe -Wl,-q 
+# RUN: llvm-bolt %t.exe --inline-memcpy -o %t.bolt 2>&1 | FileCheck %s --check-prefix=CHECK-INLINE
```
afaiu it isn't on by default, since in [`BinaryPassManager.cpp` (#L248-L253)](https://github.com/llvm/llvm-project/blob/main/bolt/lib/Rewrite/BinaryPassManager.cpp#L248-L253) the `cl::init` option isn't set for `--inline-memcpy`:

```cpp
static cl::opt<bool> StringOps("inline-memcpy",
  cl::desc("inline calls to memcpy and memmove with width-optimized "
           "load/store sequences"),
  cl::ZeroOrMore, cl::cat(BoltOptCategory));
```

Without it the tests fail

yafet-a committed:
Update conditional formatting and move check for size into binaryPasses

yafet-a commented on bolt/test/runtime/AArch64/inline-memcpy.s:
```diff
@@ -0,0 +1,193 @@
+# CHECK-ASM-NOT: bl{{.*}}<memcpy
+# CHECK-ASM-NOT: ldr{{.*}}q{{[0-9]+}}
+
+	.text
```
You're right, our search is generic (finds any instruction writing to x2 using getWrittenRegs) but our extraction implementation in AArch64MCPlusBuilder is specific (it only handles `MOVZXi`). We do find ADD/ORR etc instructions but just return nullopt for them.

I did this because ADD instruction support introduces significant complexity without the returns for implementing it. It would require some form of dataflow analysis E.g.When we encounter `add x2, x3, #4`, we cannot determine the final value without knowing `x3`'s current value. 

So in a sequence like `mov x3, #0; add x2, x3, #4`, we would need to scan backwards, build a register value map, track that x3 contains 0, then compute that x2 becomes 4, repeating this for potentially further instructions.

I can imagine this becoming more complex with register overwrites, chained ADD instructions, or cases where source register vals come from function parameters or other non-trivial sources I can't come up with rn. These would all need to be handled correctly. This more drawn out dataflow analysis would also need to be performed for every memcpy call and that might come with some overhead.

**tldr: If this support for ADD is something that would be desirable in conjunction with the support we have for MOVZXi, I am happy to look into implementing a solution for this (in this or even a separate PR). But yes, this is the reason there are no positive tests for ADD**

yafet-a committed:
Negative Tests (live-in, register move, non-mov instruction)

yafet-a commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -2597,6 +2597,132 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+
+  std::optional<uint64_t>
+  extractMoveImmediate(const MCInst &Inst, MCPhysReg TargetReg) const override {
+    if (Inst.getOpcode() == AArch64::MOVZXi && Inst.getNumOperands() >= 3) {
```
Thanks! I have looked through and removed them where appropriate

yafet-a committed:
memcpy8 redundant handling removed

yafet-a committed:
nit: comment clean up

yafet-a committed:
minor refactor

sjoerdmeijer commented on bolt/lib/Passes/BinaryPasses.cpp:
```diff
@@ -1866,8 +1866,35 @@ Error InlineMemcpy::runOnFunctions(BinaryContext &BC) {
         const bool IsMemcpy8 = (CalleeSymbol->getName() == "_memcpy8");
         const bool IsTailCall = BC.MIB->isTailCall(Inst);
 
+        // Extract size from preceding instructions (AArch64 only).
```
Can you abstract this all away? The isAArch64 check isn't really pretty and generic. I think we need a `findMemcpySizeInBytes` virtual target function or something along those lines that is a no-op for all targets, except AArch64 and this implementation should be moved to that function.

yafet-a committed:
NFC: Post-review refactor

yafet-a committed:
NFC: Test for corner case with size 0

yafet-a commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -2597,6 +2597,115 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+  }
+
+  std::optional<uint64_t>
+  extractMoveImmediate(const MCInst &Inst, MCPhysReg TargetReg) const override {
```
Agreed. Thanks for the suggestion. Done.

yafet-a committed:
Use temp instead of argument registers

yafet-a committed:
Update early return

yafet-a committed:
Update tests to be more specific about registers + negative test on early return check

yafet-a commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -2597,6 +2597,120 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+    // Generate optimal instruction sequences based on exact size.
+    switch (Size) {
+    case 1:
+      AddLoadStorePair(AArch64::LDRBBui, AArch64::STRBBui, AArch64::W3);
```
The tests have been updated to check for the exact temporary registers being used.

yafet-a committed:
Complex test + register aliasing

yafet-a commented:
> What you see there is that X9 is saved to the stack before the call to memcpy, and after the call it is reloaded because it is used by function `use`. It's a caller-saved register, and that means that if we now start using X9 as a temp register for the inlined memcpy, we are good.
> 
> Can you add this test-case to your positive tests please? You can keep the other little examples that you have, but I think it would be good to have a bigger test case where you match the whole assembly sequence that includes this caller-saved register behaviour. It would be good if you then add one more, that similarly test the whole sequence but using the FP temp register, that would then cover everything I think.

Thanks for the example. I've been looking at it this afternoon and found that it exposed my missing handling of register aliasing. Your `complex_operation` uses mov w2, #64 but our code was only looking for writes to X2, missing the 32-bit alias W2 that was actually being written to.

I initially tried using the `isMOVW()` helper, but found that it that matches MOVN, MOVK, and other variants that don't directly set the value (e.g., movk w2, #64, lsl #16 only updates part of w2) (feel free to let me know if a negative test would be needed for this). So I went with matching only the direct-assignment variants (MOVZXi/MOVZWi). Originally I was only handling MOVZXi, but your test case lmk that I needed MOVZWi too since mov w2, #64 can be emitted also.

Your test case also made me realise I needed register aliasing detection since even after matching both instruction variants, I still need `WrittenRegs.anyCommon(SizeRegAliases)` bc `getIntArgRegister(2)` returns X2, but the matched instruction might write to W2. Correct me if I am wrong but since W2 and X2 are the same physical register, I would need to detect writes to either variant.

Also I have modified your test very slightly. I changed it to 8 bytes for the integer version to properly test the X9 scenario and created a separate FP version using your integer logic as a template but with SIMD operations for the 64-byte case.

yafet-a committed:
NFC use if initializer

paschalis-mpeis reviewed (commented):
Hi @yafet-a,

Thanks for working to support memcpy! Nice patch and report. :)
Please see some comments below.

yafet-a committed:
[style] trailing whitespaces removed

yafet-a committed:
[test] CHECK-NEXT used

yafet-a committed:
[test] updated negative test to check for negative size

yafet-a committed:
[nfc] minor refactor

yafet-a committed:
[bug] memcpy call removed for sizes>64

yafet-a commented on bolt/test/runtime/AArch64/inline-memcpy.s:
```diff
@@ -0,0 +1,423 @@
+# 0-byte copy should be inlined with no load/store instructions (nothing to copy)
+# CHECK-ASM-LABEL: <test_0_byte>:
+# CHECK-ASM-NOT: ldr
+# CHECK-ASM-NOT: str  
```
Done!

yafet-a commented on bolt/test/runtime/AArch64/inline-memcpy.s:
```diff
@@ -0,0 +1,423 @@
+# CHECK-ASM-LABEL: <test_37_byte_arbitrary>:
+# CHECK-ASM: ldr{{.*}}q16, [x1]
+# CHECK-ASM: str{{.*}}q16, [x0]
+# CHECK-ASM: ldr{{.*}}q16, [x1, #0x10]
```
Yes, this is a good suggestion. It's been updated now.

yafet-a commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -2597,6 +2597,121 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+      WrittenRegs.reset();
+      getWrittenRegs(Inst, WrittenRegs);
+
+      if (SizeReg != getNoRegister() && WrittenRegs.anyCommon(SizeRegAliases);
```
Yes, you're right. I have moved it out of the loop now. I have also moved the call to WrittenRegs to be after the check for the size register too since that'd be slightly more efficient.

yafet-a commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -2597,6 +2597,121 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+
+  InstructionListType createInlineMemcpy(bool ReturnEnd) const override {
+    return createInlineMemcpy(ReturnEnd, std::nullopt);
+  }
```
It isn't needed you're correct, I kept it here initially as I had thought it might be clearer but the overloading of the arguments should be clear enough. Thanks for the suggestion, it has been removed.

paschalis-mpeis reviewed (approved):
Thanks for addressing the comments, @yafet-a.

Looks good. Tentative accept, pending the verification ask from @sjoerdmeijer and also give it a few days for the rest of the reviewers to react.

Also I added a couple of nits.

yafet-a commented:
> Did you check whether some of the inlined memcpy's were actually executed? I want to avoid that we have missed some corner cases.

Yes, I checked on the zstd smoke tests and inspected the objdump. Found our inlined memcpy patterns at multiple addresses:

```assembly
178bc: ldr q0, [x1]
178c0: str q0, [x0]

40e4c: ldr q0, [x1]
40e50: str q0, [x0]

46e10: ldr q0, [x1]
46e14: str q0, [x0]
```

I then confirmed execution by using perf tools on the `ZSTD_compressBlock_fast` function and saw some of these instances actively running e.g.:

```assembly
0.33 :   559bc:  ldr     q0, [x1]    <- 0.33% CPU time 
   |         559c0:  str     q0, [x0]    <- actively executing
```

So for this 16-byte copy example it used our inlining change that replaced the original memcpy calls.
<img width="567" height="146" alt="image" src="https://github.com/user-attachments/assets/878545ad-8393-476c-b693-f8f62847184a" />

yafet-a committed:
[nfc][test] reordered test

yafet-a committed:
[nfc] added assert for default case (future-proofing for changes to BinaryPasses.cpp)

yafet-a commented on bolt/test/runtime/AArch64/inline-memcpy.s:
```diff
@@ -0,0 +1,384 @@
+	bl	memcpy
+	ldp	x29, x30, [sp], #32
+	ret
+	.size	test_negative_size, .-test_negative_size
```
Done, thanks!

sjoerdmeijer reviewed (approved):
Thanks for the testing, LGTM.

Let's wait a day with merging this in case someone else has more comments.
